### ISSUE TITLE
Critical error - OutOfMemoryError with WebFlux when DEBUG enabled 

### ISSUE THREAD

**Issue by asarkar** [spring-projects/spring-integration#2309]
I've an app where a large file is streamed from a Web server to Couchbase server using Spring Integration 5 WebFlux support. The file size is about 4GB compressed, and 12GB uncompressed. The app works as expected when logging level is set to INFO or above, but when DEBUG is enabled, crashes with `OutOfMemoryError`. There are no log statements in the streaming flow that are written by me, so I'm convinced that this is a problem with how Spring Integration handles `Flux`. Also, funny thing is, I don't see additional log statements when DEBUG enabled, the app just sits there until it runs out of memory.

Following are the  relevant code snippets. I'm using [Spring Boot logging](https://docs.spring.io/spring-boot/docs/current/reference/html/howto-logging.html) support.
```
@EnableIntegrationManagement(
        defaultLoggingEnabled = "false"
)
``` 

```
@Bean
public IntegrationFlow artifactStreamingFlow() {
    // Direct channel
    return IntegrationFlows.from(artifactoryProperties.getArtifactStreamingInputChannel())
            .filter(Message.class, msg -> msg.getHeaders().containsKey(FILE_TYPE_HEADER))
            .handle(new GzipToFluxTransformer(artifactoryProperties))
            .transform((Flux<String> payload) -> payload
                    .onBackpressureBuffer(artifactoryProperties.getOnBackpressureBufferSize(),
                            s -> log.error("Buffer overrun!")))
            .split()
            // A FluxMessageChannel is created for this
            .channel(c -> c.flux(artifactoryProperties.getArtifactStreamingOutputChannel()))
            .get();
}
```
```
@Bean
public IntegrationFlow aggregatingFlow() {
    // Output channel from the artifactStreamingFlow
    return IntegrationFlows.from(publisher)
            .enrichHeaders(spec -> spec.header(ERROR_CHANNEL, SINK_ERROR_CHANNEL))
            .filter(Message.class, msg -> msg.getHeaders().containsKey(FILENAME_HEADER))
            .aggregate(spec -> spec
                    .correlationStrategy(new HeaderAttributeCorrelationStrategy(FILENAME_HEADER))
                    .releaseStrategy(new MessageCountReleaseStrategy(couchbaseProperties.getBatchSize()))
                    .expireGroupsUponCompletion(true) // allows new groups with same correlation id
                    .expireGroupsUponTimeout(true) // same as above
                    .groupTimeout(couchbaseProperties.getBatchTimeoutMillis())
                    .sendPartialResultOnExpiry(true) // send whatever is accumulated if group times out
            )
            // An ExecutorChannel is created for this
            .channel(couchbaseProperties.getArtifactAggregatingOutputChannel())
            .get();
}
```
```
@Bean
public IntegrationFlow artifactLoadingFlow() {
     // Output channel from the aggregatingFlow
    return IntegrationFlows.from(couchbaseProperties.getRepositoryInputChannel())
            .transform(Collection.class, repository::saveAll)
            .transform((Observable<RawJsonDocument> o) -> {
                o.subscribe(doc -> {
                        },
                        t -> log.error("Failed to persist document.", t));
                return EMPTY;
            })
            // IntegrationContextUtils.NULL_CHANNEL_BEAN_NAME
            .channel(couchbaseProperties.getRepositoryOutputChannel())
            .get();
}
```
```
@Bean(ARTIFACT_STREAMING_OUTPUT_CHANNEL)
public MessageChannel artifactStreamingOutputChannel() {
    return new FluxMessageChannel();
}

@Bean(ARTIFACT_AGGREGATING_OUTPUT_CHANNEL)
public MessageChannel artifactAggregatingOutputChannel() {
    SimpleAsyncTaskExecutor executor = new SimpleAsyncTaskExecutor("yelp-sink-");
    executor.setConcurrencyLimit(couchbaseProperties.getMaxConcurrency());

    return new ExecutorChannel(executor);
}
```

**Comment by garyrussell** [comment]
Strange - hard to see how log level would affect this, especially given that you've disabled all logging by Spring Integration components.

To track this down, I think we'll need a complete example (small as possible - eliminating as many external deps as possible - couchbase etc) that exhibits the behavior.

**Comment by asarkar** [comment]
I'm afraid reproducing this in a minimal application won't be easy, if even possible. I've verified with a small input file (< 500k), and the code does work even with DEBUG enabled. It appears to me that somewhere in the code, the stream is buffered in memory, which for small file sizes, isn't an issue, but for bigger files, causes the OOM. I would also like to change fewer variables (like removing Couchbase), since we don't know which one may be causing the problem.

I'm suggesting taking the original code with as little modification as possible, and run it with a Couchbase community edition Docker container as a sink. The source is a [JFrog Artifactory](https://jfrog.com/artifactory/), that offers free trials, so I can use a Docker container there too. That way, we still keep the same architecture, but reduce the setup to a few `docker run` commands.

If the above suggestion is acceptable, then I need to figure out how to strip off any proprietary stuff and come up with a reproducible solution.

**Comment by garyrussell** [comment]
>I would also like to change fewer variables (like removing Couchbase), since we don't know which one may be causing the problem.

Unfortunately, if you can't eliminate third party libraries, it's doubtful we can help. I hope you understand that we can only help with debugging our code.

Why can you not reproduce with a similar file loaded from a local file system instead?

**Comment by asarkar** [comment]
> similar file loaded from a local file system

I'll try to repro using an Nginx source and writing to local FS instead of Couchbase. That way, I can keep the HTTP streaming source, just change the sink.

**Comment by garyrussell** [comment]
Sure, sounds good.

**Comment by asarkar** [comment]
I've been able to find the root cause of the issue, and it is not Spring integration. Attempting to reduce the project to the minimum helped me narrow down the problem. It was caused by an OkHttp logging interceptor that I was using. I've filed a bug against that project, and also found an alternative for it.
https://github.com/devcsrj/slf4j-okhttp3-logging-interceptor/issues/3

Thanks for indulging me, and I apologize for the distraction.
